### ISSUE TITLE
Ignore encryption data in mdat if senc box is present

### ISSUE THREAD

**Issue by sandersaares** [google/ExoPlayer#861]
We are moving onward toward our goals to achieve playback of a single video on many players, with the last encounter described here https://github.com/google/ExoPlayer/issues/835 with some apparent resolution for our issue having been found.

Putting the encryption data into the `mdat` box and making sure there is no extra data appeared to be a good solution last week. However, in the intervening period, we have found at least one player that is unable to read the encryption data from `mdat` and requires `senc` to be present.

The next step for us was to use both the `senc` box and the `mdat` box with the `saio` offset pointing into the `mdat`. The idea being that if the player does not understand `senc` it will just read the data out of the `mdat` as instructed by `saio` and if it does require `senc` it will just use that. It feels incredibly dirty but it just about works... except with ExoPlayer!

Video URL: http://media.axprod.net/ExoPlayer/Multisdk_SencAndMdat/Encrypted/manifest.mpd
Widevine license server URL: http://axpr-wv-fe.cloudapp.net:8080/LicensingService

From the log, I see that it tries to read the start of the `mdat` as samples, presumably because it encountered `senc` and thus does not know that there is going to be more encryption data in `mdat`. At least this is my current theory - have not actually attached a debugger but from reading the log, it seems straightforward enough.

```
10-09 16:10:59.670: E/LoadTask(3293): java.lang.IllegalStateException: Top bit not zero: -1006207405
10-09 16:10:59.670: E/LoadTask(3293): 	at com.google.android.exoplayer.util.ParsableByteArray.readUnsignedIntToInt(ParsableByteArray.java:240)
10-09 16:10:59.670: E/LoadTask(3293): 	at com.google.android.exoplayer.extractor.mp4.FragmentedMp4Extractor.readSample(FragmentedMp4Extractor.java:658)
10-09 16:10:59.670: E/LoadTask(3293): 	at com.google.android.exoplayer.extractor.mp4.FragmentedMp4Extractor.read(FragmentedMp4Extractor.java:170)
10-09 16:10:59.670: E/LoadTask(3293): 	at com.google.android.exoplayer.chunk.ChunkExtractorWrapper.read(ChunkExtractorWrapper.java:93)
10-09 16:10:59.670: E/LoadTask(3293): 	at com.google.android.exoplayer.chunk.ContainerMediaChunk.load(ContainerMediaChunk.java:165)
```

I would thus request a change to ExoPlayer to make it ignore this encryption data inside `mdat` even if it is already parsed from `senc` (detected by checking whether `saio` points into `mdat`).

**Comment by ojw28** [comment]
So what happens here is that we parse what we need from the `senc` box, and hence (currently) expect the `mdat` box to only contain sample data. At this point the encryption data in the `mdat` is, as far as we're concerned, miscellaneous data that we don't care about. Given this, I'm pretty sure the fix for #837 will also fix this issue, since that issue is tracking exactly "handle miscellaneous data that we don't care about in the mdat box". Does that make sense?

**Comment by sandersaares** [comment]
Yes, I agree with your analysis that a fix for https://github.com/google/ExoPlayer/issues/837 would also solve this.
